waiter/src/waiter/auth/oauth/github.clj

```diff
@@ -0,0 +1,89 @@
+            [qbits.jet.client.http :as http]
+            [ring.middleware.params :as ring-params]
+            [ring.util.codec :as ring-codec]
+            [waiter.auth.authentication :as auth]
+            [waiter.auth.oauth :as oauth]
+            [waiter.cookie-support :as cookie-support]))
+
+; See https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/identifying-users-for-github-apps/
+(defrecord GitHubOAuthProvider [authenticate-uri-fn client-id client-secret http-client password]
+  oauth/OAuthProvider
+  (display-name [_] "GitHub")
+  (redirect [_ {:keys [query-string] {:keys [path]} :route-params :as request}]
+    (fa/go-try
+      (let [token (oauth/random-str)
+            state (-> {:location (oauth/make-location {:path path
+                                                       :query-string query-string})
+                       :token token}
+                      json/write-str)
+            location (oauth/make-uri "https://github.com/login/oauth/authorize"
+                                     {"client_id" client-id
+                                      "scope" "user:email"
+                                      "redirect_uri" (authenticate-uri-fn request)
+                                      "state" state})]
+        (-> {:status 307
+             :headers {"location" location}}
+            ;; We should store this cookie only long enough to provide a reasonable amount of time
+            ;; for the client to authenticate.
+            (cookie-support/add-encoded-cookie password oauth/OAUTH-COOKIE-NAME token (-> 15 t/minutes t/in-seconds))))))
+  (authenticate [_ request]
+    (fa/go-try
+      (let [{:strs [code state] :as params} (:params (ring-params/params-request request))
+            _ (when-not (and code state)
+                (throw (ex-info "Malformed request" {:status 400
+                                                     :params params})))
+            {:strs [location token]} (try (-> state
+                                              json/read-str)
+                                          (catch Exception e
+                                            (throw (ex-info "Couldn't parse state from GitHub" {:status 403} e))))
+            _ (oauth/verify-oauth-cookie request token password)
+            {:keys [body error status]} (async/<! (http/post http-client "https://github.com/login/oauth/access_token"
+                                                             {:form-params {"client_id" client-id
+                                                                            "client_secret" client-secret
+                                                                            "code" code
+                                                                            "state" state}}))
+            _ (when error
+                (throw (ex-info "Error while communicating with GitHub" {:status 403} error)))
+            _ (when-not (= status 200)
```

Comment:
I did not find where the [github doc](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/identifying-users-for-github-apps/) says to expect only a 200 response for success.

Reply:
Which additional status codes do we think would also represent a valid response?  Do we think `204 No Content` would be valid?

Reply:
I would vote for checking 2XX  `(<= 200 status 299)`

Reply:
At the very least we should log why the auth failed.